common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+    uint32_t bit = 1;
+    for (ssize_t i : llvm::seq<ssize_t>(0, Size)) {
+      if (bytes[i] == Deleted) {
+        return MatchIndex(bit);
+      }
+      bit <<= 1;
+    }
+    return MatchIndex(0);
+  }
+
+  // Materialize the group into a word.
+  uint64_t mask = byte_ints[0] | (~byte_ints[0] << 7);
+  mask = ~mask & MSBs;
+
+  CARBON_DCHECK(
+      VerifyIndexMask(mask, [](uint8_t byte) { return byte == Deleted; }));
+  return MatchIndex(mask);
+}
+
+inline auto MetadataGroup::PortableCompareEqual(MetadataGroup lhs,
+                                                MetadataGroup rhs) -> bool {
+  return llvm::equal(lhs.bytes, rhs.bytes);
+}
+
+inline auto MetadataGroup::SIMDLoad(const uint8_t* metadata, ssize_t index)
+    -> MetadataGroup {
+  MetadataGroup g;
+#if CARBON_NEON_SIMD_SUPPORT
+  g.byte_vec = vld1_u8(metadata + index);
+#elif CARBON_X86_SIMD_SUPPORT
+  g.byte_vec =
+      _mm_load_si128(reinterpret_cast<const __m128i*>(metadata + index));
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+  return g;
+}
+
+inline auto MetadataGroup::SIMDStore(uint8_t* metadata, ssize_t index) const
+    -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  vst1_u8(metadata + index, byte_vec);
+#elif CARBON_X86_SIMD_SUPPORT
+  _mm_store_si128(reinterpret_cast<__m128i*>(metadata + index), byte_vec);
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+}
+
+inline auto MetadataGroup::SIMDClearDeleted() -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  byte_ints[0] &= (~LSBs | byte_ints[0] >> 7);
+#elif CARBON_X86_SIMD_SUPPORT
+  byte_vec = _mm_blendv_epi8(_mm_setzero_si128(), byte_vec, byte_vec);
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+}
+
+inline auto MetadataGroup::SIMDMatch(uint8_t present_byte) const -> MatchRange {
+  MatchRange result;
+#if CARBON_NEON_SIMD_SUPPORT
+  auto match_byte_vec = vdup_n_u8(present_byte | PresentMask);
+  auto match_byte_cmp_vec = vceq_u8(byte_vec, match_byte_vec);
+  uint64_t mask = vreinterpret_u64_u8(match_byte_cmp_vec)[0];
+  // Mask in scalar to fold with testing for zero.
+  result = MatchRange(mask & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  result = X86SIMDMatch(present_byte | PresentMask);
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(present_byte);
+#endif
+  return result;
+}
+
+inline auto MetadataGroup::SIMDMatchPresent() const -> MatchRange {
+  MatchRange result;
+#if CARBON_NEON_SIMD_SUPPORT
+  // Just extract directly.
+  uint64_t mask = vreinterpret_u64_u8(byte_vec)[0];
+  // Mask in scalar to fold with testing for zero.
+  result = MatchRange(mask & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  // We arrange the byte vector for present bytes so that we can directly
+  // extract it as a mask.
+  result = MatchRange(_mm_movemask_epi8(byte_vec));
```

Comment:
Not sure what this operation does, and so why you don't need to `& MSBs` like the other code paths.

Reply:
This produces a bit-encoded mask instead of a byte-encoded mask.

This is also probably the source of me calling the backing integer storage for BitIndex and BitIndexRange a `mask` of some kind -- that's how x86's SIMD instructions refer to them.

Reply:
That sounds worth including in the comment.

Reply:
Tweaked the comment a bit, but to an extent I think this is just the documented contract of `_mm_movemask_epi8`.

I have added the links to the reference manuals for SIMD extensions to the top of the file. A bit tricky to link to a single intrinsic, but I think this will work: https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=movemask&ig_expand=4602